src/mon/OSDMonitor.cc

```diff
@@ -2983,9 +2983,14 @@ void OSDMonitor::get_health(list<pair<health_status_t,string> >& summary,
   } else {
     int num_in_osds = 0;
     int num_down_in_osds = 0;
+    set<int> osds;
+    osdmap.crush->get_all_osds_in_crush(osds);
     for (int i = 0; i < osdmap.get_max_osd(); i++) {
-      if (!osdmap.exists(i) || osdmap.is_out(i))
-	continue;
+      if (!osdmap.exists(i))
+	 continue;
+      osds.erase(i);
```

Comment:
Instead of generating set of OSDs from name_map then erasing most of them, you could just check if they're in name_map and then add them to the set if they're not in osdmap.